lib/spack/spack/spec.py

```diff
@@ -1551,6 +1551,18 @@ def concretize(self):
         # Mark everything in the spec as concrete, as well.
         self._mark_concrete()
 
+        # Now that the spec is concrete we should check if
+        # there are declared conflicts
+        matches = []
+        for x in self.traverse():
+            for conflict_spec, when_list in x.package.conflicts.items():
+                if x.satisfies(conflict_spec):
+                    for when_spec in when_list:
+                        if x.satisfies(when_spec):
+                            matches.append((x, conflict_spec, when_spec))
+        if matches:
```

Comment:
@davydden The lines above (from the for loop) are the logic behind the `conflicts` directive. Basically with conflicts you store a map that has the conflict as the key, and a list of conditions as values. This allows you to use the same conflict multiple times:
```python
# +foo cannot be used with intel
conflicts('+foo', when='%intel') 
# or when we link with zlib
conflicts('+foo', when='^zlib')
```

If you satisfy the conflict and at least one condition the operation is aborted.